### PR TITLE
RLM-54 Fix ES upgrade log in leapfrog

### PR DESCRIPTION
**Pull request by Frank-ZhangXin**
ES container has been removed from leapfrog's container destroy list,
and ES (re)deployment also would be avoided in leap ugprade as well.
Every ES ugprade operation is supposed to be performed after all leap
done. (Next to-do: automate ES upgrade operation into current leap)

Issue: [RLM-54](https://rpc-openstack.atlassian.net/browse/RLM-54)

### CODE REVIEW

File: scripts/leapfrog/post_leap.sh [hunk 1]
```diff
@@ -29,6 +29,8 @@ if [[ ! -f "${UPGRADE_LEAP_MARKER_FOLDER}/deploy-rpc.complete" ]]; then
     # TODO(remove the following hack to restart the neutron agents, when fixed upstream)
     ansible -m shell -a "restart neutron-linuxbridge-agent" nova_compute -i /opt/rpc-openstack/openstack-ansible/playbooks/inventory/dynamic_inventory.py
     openstack-ansible ${RPCO_DEFAULT_FOLDER}/scripts/leapfrog/playbooks/remove-old-agents-from-maas.yml
+    # TO-DO(fran0021) exclude Elasticsearch, upgrade it after leap
```
Comment (claco) [inline scripts/leapfrog/post_leap.sh]: @Frank-ZhangXin @antonym @cloudnull This will also effect K2N leapfrogs should we find ourselves wanting to test/bump future ones to say r14.5 or so after this weekend and the next few upgrades right?

Is there a way to make this L+ only?

Comment (antonym) [reply scripts/leapfrog/post_leap.sh]: The pre and post parts in rpc-o of the leapfrog code don't currently have version detection.  We might see if osa_ops drops the leapfrog from variable down anywhere and if not, add it so that we can then key off it in the post scripts.

Comment (claco) [reply scripts/leapfrog/post_leap.sh]: IT may also be that @cloudnull rcbops/rpc-upgrades tackles this via getting these changes there, and using that tool for L2N?

Comment (antonym) [reply scripts/leapfrog/post_leap.sh]: That's ultimately where those changes would live but as it is now, the rpc-upgrades/leapfrog scripts in rpc-o only contain the wrapper around the openstack-ansible-ops leapfrog logic.  I'm not sure if that's something we want to fork or leave upstream.  Either way, it looks like the changes need to happen at the end before we run the rpc playbooks.  

We can probably load up a variable there that specifies what the original version was that was upgraded and just set a if != kilo, then run.

**Comment by major** [comment]
recheck_cit_trusty_swift_deploy

File: rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml [hunk 2]
```diff
@@ -231,7 +231,7 @@
     - elasticsearch-reindex
 
 - name: Monitor reindexing process
-  shell: /opt/reindex-liberty.py --host "{{ hostvars[groups['elasticsearch_container'][0]]['container_address'] }}" --monitor --continuous
+  shell: /opt/reindex-liberty.py --host "{{ hostvars[groups['elasticsearch_container'][0]]['container_address'] }}" --monitor --continuous --verbose
```
Comment (prometheanfire) [inline rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml]: I don't see a reason in the commit message for this change (adding `--verbose`).  If this is actually needed, please note why in the commit message.

Comment (claco) [reply rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml]: Also, side related, I want to say we needed to tweak --host here to or something. @Frank-ZhangXin mentuioned it in the comments of that card as well.

**Comment by claco** [comment]
@Frank-ZhangXin  Is --host also still needed from here?

https://rpc-openstack.atlassian.net/browse/RLM-54?focusedCommentId=170522&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170522

File: scripts/functions.sh [hunk 3]
```diff
@@ -58,6 +58,9 @@ export HOST_RCBOPS_REPO=${HOST_RCBOPS_REPO:-"http://rpc-repo.rackspace.com"}
 # Derive the rpc_release version from the group vars
 export RPC_RELEASE="$(/opt/rpc-openstack/scripts/artifacts-building/derive-artifact-version.sh)"
 
+# Upgrade of Elasticsearch
+export UPGRADE_ELASTCSEARCH=${UPGRADE_ELASTICSEARCH:-"no"}
```
Comment (antonym) [inline scripts/functions.sh]: Variable misspelled.

File: rpcd/playbooks/setup-logging.yml [hunk 4]
```diff
@@ -14,6 +14,7 @@
 # limitations under the License.
 
 - include: elasticsearch.yml
+  when: "{{ lookup('env', 'UPGRADE_ELASTICSEARCH')}}" == 'no'
```
Comment (antonym) [inline rpcd/playbooks/setup-logging.yml]: Travis job isn't too happy with this lookup: https://travis-ci.org/rcbops/rpc-openstack/jobs/292866233#L762

File: scripts/functions.sh (hunk 3 above)
Comment (antonym) [inline scripts/functions.sh]: Misspelled variable

File: rpcd/playbooks/setup-logging.yml (hunk 4 above)
Comment (antonym) [inline rpcd/playbooks/setup-logging.yml]: Travis job isn't too happy with this lookup: https://travis-ci.org/rcbops/rpc-openstack/jobs/292866233#L762

File: rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml (hunk 2 above)
Comment (Frank-ZhangXin) [reply rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml]: @prometheanfire
"--verbose" is not strictly required here since ansible won't print out monitoring progress in the play, so it can be removed.
@claco
"--host" arg is required if any es operation is asked to be performed.

File: rpcd/playbooks/setup-logging.yml (hunk 4 above)
Comment (Frank-ZhangXin) [reply rpcd/playbooks/setup-logging.yml]: @antonym is it supposed to be down below?
when: "{{ lookup('env', 'UPGRADE_ELASTICSEARCH')}}": 'no'

**Comment by Frank-ZhangXin** [comment]
@claco yes, the old error you linked is caused by missing '--host' arg.

File: rpcd/playbooks/setup-logging.yml [hunk 5]
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 
 - include: elasticsearch.yml
+  when:
+    - lookup('env', 'UPGRADE_ELASTICSEARCH') != "yes"
```
Comment (major) [inline rpcd/playbooks/setup-logging.yml]: This is a little more readable:

```
  when:
    - not lookup('env', 'UPGRADE_ELASTICSEARCH') | bool
```

File: scripts/leapfrog/elasticsearch_upgrade.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+# Run elasticsearch upgrade playbook
+pushd /opt/rpc-openstack/rpcd/playbooks
+openstack-ansible -e 'logging_upgrade=true' --tags 'reindex-wrapper,elasticsearch-upgrade' elasticsearch.yml
+popd
+
+# Restore logstash service on Logstash container
+pushd /opt/rpc-openstack/rpcd/playbooks
+openstack-ansible ../../scripts/leapfrog/playbooks/logstash-start.yml
+popd
```
Comment (major) [inline scripts/leapfrog/elasticsearch_upgrade.sh]: Is there a reason for the multiple `pushd`/`popd` lines here? Could we just put all of these lines together?

Comment (Frank-ZhangXin) [reply scripts/leapfrog/elasticsearch_upgrade.sh]: Just want to make things uniform, but I can make them look simpler.

File: rpcd/playbooks/setup-logging.yml (hunk 5 above)
Comment (Frank-ZhangXin) [reply rpcd/playbooks/setup-logging.yml]: Yeah, good idea. Thanks!

**Comment by Frank-ZhangXin** [comment]
Update:
We're about to move this upgrade along with all leapfrog update in to RPC upgrade soon.

File: rpcd/playbooks/setup-logging.yml [hunk 7]
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 
 - include: elasticsearch.yml
+  when:
+    - not lookup('env', 'UPGRADE_ELASTICSEARCH') | bool
```
Comment (cloudnull) [inline rpcd/playbooks/setup-logging.yml]: this variable needs something in docs so users know what this is and why they might want to set it.

File: scripts/functions.sh [hunk 8]
```diff
@@ -58,6 +58,9 @@ export HOST_RCBOPS_REPO=${HOST_RCBOPS_REPO:-"http://rpc-repo.rackspace.com"}
 # Derive the rpc_release version from the group vars
 export RPC_RELEASE="$(/opt/rpc-openstack/scripts/artifacts-building/derive-artifact-version.sh)"
 
+# Upgrade of Elasticsearch
+export UPGRADE_ELASTICSEARCH=${UPGRADE_ELASTICSEARCH:-"no"}
```
Comment (cloudnull) [inline scripts/functions.sh]: what is this and why is this here?

File: scripts/leapfrog/elasticsearch_upgrade.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+## Shell Opts ----------------------------------------------------------------
+set -e -u -x
+set -o pipefail
+
+# Check upgrade flag
+if [[ -n "$UPGRADE_ELASTICSEARCH" ]]; then
+  if [[ "$UPGRADE_ELASTICSEARCH" == "NO" ]]; then
```
Comment (cloudnull) [inline scripts/leapfrog/elasticsearch_upgrade.sh]: The default in the above variable is "no" should the above default option be "NO"?

Also with the casting of strings we might want to consider making this `true or `false`. 

something like:

``` bash
if [ ${UPGRADE_ELASTICSEARCH} = false ]; then
...
fi
```

File: scripts/leapfrog/elasticsearch_upgrade.sh [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    exit
+  fi
+else
+  echo "Please preset env var before leap."
+  exit
+fi
+
+# Update pip.conf on Elasticsearch container
+pushd /opt/rpc-openstack/rpcd/playbooks
+openstack-ansible ../../scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml
```
Comment (cloudnull) [inline scripts/leapfrog/elasticsearch_upgrade.sh]: For consistency and readability these commands should all be indented two spaces between the pushd/popd lines.

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+- name: Upgrade pip on Elasticsearch container
+  hosts: elasticsearch_all
+  user: root
+  roles:
+    - role: "pip_install"
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: Why do we need the pip_install role run within an existing elk container and why does this need to be done post-leap? shouldn't our elk role take care of this already?

File: scripts/leapfrog/playbooks/logstash-start.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/logstash-start.yml]: This playbook needs the boiler plate license for RPC-OpenStack

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml [hunk 13]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: This playbook needs the boiler plate license for RPC-OpenStack

File: scripts/leapfrog/playbooks/logstash-start.yml [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+- name: Start logstash serivce
+  hosts: logstash_all
+  user: root
+  tasks:
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/logstash-start.yml]: Should this be some form a rolling restart or can we really start all logstash services at once?

File: scripts/leapfrog/playbooks/logstash-start.yml [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+    - name: Service start
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/logstash-start.yml]: Is this not covered by the elk upgrade process?

File: scripts/leapfrog/playbooks/logstash-stop.yml [hunk 16]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Stop logstash serivce
+  hosts: logstash_all
+  user: root
+  tasks:
+    - name: Service stop
+      service:
```
Comment (cloudnull) [inline scripts/leapfrog/playbooks/logstash-stop.yml]: Is this not covered by the existing elk upgrade process?

File: rpcd/playbooks/setup-logging.yml (hunk 7 above)
Comment (Frank-ZhangXin) [reply rpcd/playbooks/setup-logging.yml]: Yeah, I plan to wrap all the instruction and docs when we approve this.

File: scripts/functions.sh (hunk 8 above)
Comment (Frank-ZhangXin) [reply scripts/functions.sh]: By default deployment, the upgrade flag is supposed to be negative. My intending is wishing user to manually export this flag to true.

File: scripts/leapfrog/elasticsearch_upgrade.sh (hunk 9 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/elasticsearch_upgrade.sh]: yeah, I was confused with all existing env var we're using. Sometime it is true/false, and sometimes it is yes/no. I'll match the var value in default here, I mean match 'no' to 'NO'.

File: scripts/leapfrog/elasticsearch_upgrade.sh (hunk 10 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/elasticsearch_upgrade.sh]: Like?
  pushd xxx
openstack-ansible xxx
  popd 
I can fix it.

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml (hunk 13 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: yes, you're right, adding it.

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml (hunk 11 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: The fact is the way to upgrade elasticsearch require us keeping old elasticsearch container. Well the old container won't match pip config which elasticsearch upgrade playbook asked. The most straightforward solution is simply running pip role on elasticsreach container.

File: scripts/leapfrog/playbooks/logstash-start.yml (hunk 12 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/logstash-start.yml]: cool, I'll add it

File: scripts/leapfrog/playbooks/logstash-start.yml (hunk 14 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/logstash-start.yml]: So you want to add more script to examine if all logstash containers service are back online?

File: scripts/leapfrog/elasticsearch_upgrade.sh (hunk 10 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/elasticsearch_upgrade.sh]: Cool will add them

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml (hunk 13 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: yes, you're right, adding it.

File: scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml (hunk 11 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/elasticsearch-postleap-pip-upgrade.yml]: The fact is the way to upgrade elasticsearch requires us keeping old elasticsearch container. Well the old container won't match pip config which elasticsearch upgrade playbook asked. The most straightforward solution is simply running pip role on elasticsreach container.

File: scripts/leapfrog/playbooks/logstash-start.yml (hunk 14 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/logstash-start.yml]: cool, I'll add it

File: scripts/leapfrog/playbooks/logstash-start.yml (hunk 15 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/logstash-start.yml]: At this point logstash has been already upgraded. The reason to stop and restart logstash is elasticsearch upgrade won't need more coming log when the re-indexing is going. So try not to forever hanging re-indexing process, we should turn off logstash temporarily and give it back when elasticsearch upgrade done.

File: scripts/leapfrog/playbooks/logstash-stop.yml (hunk 16 above)
Comment (Frank-ZhangXin) [reply scripts/leapfrog/playbooks/logstash-stop.yml]: Same like what I said above.

File: rpcd/playbooks/setup-logging.yml (hunk 7 above)
Comment (cloudnull) [reply rpcd/playbooks/setup-logging.yml]: Still needs docs

File: scripts/functions.sh (hunk 8 above)
Comment (cloudnull) [reply scripts/functions.sh]: doc this

File: releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml [hunk 17]
```diff
@@ -0,0 +1,16 @@
+---
+issues:
+  - |
+    Elasticserach log data would be lost in the process of leapgfrog upgrade 
+    due to rebuilding ELK containers. (Liberty to Newton)
+upgrade:
+  - |
+    To preserve Elasticsearch log data in the leapfrog upgrade. Two new
+    environment variables need to set. New environment variable 
+    'UGPRADE_ELASTICSEARCH' is 'no' by default. When 'UPGRADE_ELASTICSEARCH'
```
Comment (antonym) [inline releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml]: UGPRADE_ELASTICSEARCH is mispelled.

File: releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+  - |
+    Elasticserach log data would be lost in the process of leapgfrog upgrade 
+    due to rebuilding ELK containers. (Liberty to Newton)
+upgrade:
+  - |
+    To preserve Elasticsearch log data in the leapfrog upgrade. Two new
+    environment variables need to set. New environment variable 
+    'UGPRADE_ELASTICSEARCH' is 'no' by default. When 'UPGRADE_ELASTICSEARCH' 
+    is marked to 'yes', the setup of Elasticsearch 'elasticsearch.yml' play 
+    will be skipped in 'setup-logging.yml' play. Another new enironment
```
Comment (antonym) [inline releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml]: environment (sp)

File: releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    due to rebuilding ELK containers. (Liberty to Newton)
+upgrade:
+  - |
+    To preserve Elasticsearch log data in the leapfrog upgrade. Two new
+    environment variables need to set. New environment variable 
+    'UGPRADE_ELASTICSEARCH' is 'no' by default. When 'UPGRADE_ELASTICSEARCH' 
+    is marked to 'yes', the setup of Elasticsearch 'elasticsearch.yml' play 
+    will be skipped in 'setup-logging.yml' play. Another new enironment 
+    variable 'CONTAINERS_TO_DESTROY' needs to be set prior leap upgrade as 
+    well. User should "exports CONTAINERS_TO_DESTROY='all_containers
```
Comment (antonym) [inline releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml]: export (sp)

File: releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml [hunk 20]
```diff
@@ -0,0 +1,16 @@
+---
+issues:
+  - |
+    Elasticserach log data would be lost in the process of leapgfrog upgrade 
+    due to rebuilding ELK containers. (Liberty to Newton)
+upgrade:
+  - |
+    To preserve Elasticsearch log data in the leapfrog upgrade. Two new
```
Comment (antonym) [inline releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml]: upgrade,  two

File: releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml [hunk 21]
```diff
@@ -0,0 +1,16 @@
+---
+issues:
+  - |
+    Elasticserach log data would be lost in the process of leapgfrog upgrade
```
Comment (antonym) [inline releasenotes/notes/elasticsearch-upgrade-leapfrog-4f7a2f3e1f5e957d.yaml]: Elasticsearch (sp)

**Comment by Frank-ZhangXin** [comment]
@antonym @cloudnull fixed the miss-spell.

**Comment by Frank-ZhangXin** [comment]
Close this PR since we move all these stuff into RPC-upgrades repo: https://github.com/rcbops/rpc-upgrades
